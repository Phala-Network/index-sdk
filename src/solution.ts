import {u8aToHex} from '@polkadot/util'
import Ajv, {JSONSchemaType} from 'ajv'
import * as $ from 'subshape'
import {Chain, Hex, Solution, Step} from './types'

const $single = $.object(
  $.field('exe', $.str),
  $.field('sourceChain', $.str),
  $.field('destChain', $.str),
  $.field('spendAsset', $.str),
  $.field('receiveAsset', $.str)
)

const $batch = $.array($single)

interface Single extends Step {
  _tag: 'single'
}

class Batch extends Array<Step> {
  _tag = 'batch' as const
}

const $multiStep = $.taggedUnion('_tag', [
  $.variant('single', $single),
  $.variant('batch', $batch),
])

const $legacySolution = $.array($single)

const $solution = $.array($multiStep)

export const processSolution = (
  chainMap: Map<string, Chain>,
  solution: Solution
): Hex => {
  // FIXME: remove legacy solution when https://github.com/Phala-Network/index-contract/pull/92 is merged
  return u8aToHex($legacySolution.encode(solution))
  // const mergedSteps: Array<Single | Batch> = []
  // let batch: Batch = new Batch()

  // const pushSingleOrBatch = (item: Step | Batch) => {
  //   if (Array.isArray(item)) {
  //     if (item.length === 0) return
  //     if (item.length === 1) {
  //       item = item[0]
  //     } else {
  //       mergedSteps.push(item)
  //       return
  //     }
  //   }
  //   mergedSteps.push({_tag: 'single', ...item})
  // }

  // for (const step of solution) {
  //   const chain = chainMap.get(step.sourceChain) as Chain

  //   const isEvm = chain.chainType === 'Evm'
  //   const isSameChain =
  //     batch.length > 0 &&
  //     step.sourceChain === batch[batch.length - 1].sourceChain

  //   if (isEvm && (batch.length === 0 || isSameChain)) {
  //     batch.push(step)
  //   } else {
  //     pushSingleOrBatch(batch)
  //     batch = new Batch()
  //     if (isEvm) {
  //       batch.push(step)
  //     } else {
  //       pushSingleOrBatch(step)
  //     }
  //   }
  // }

  // pushSingleOrBatch(batch)

  // return u8aToHex($solution.encode(mergedSteps))
}

const hexPattern = '^0x[0-9a-fA-F]+$'

export const createValidateFn = (chains: Chain[]) => {
  const ajv = new Ajv()
  const chainName = chains.map((c) => c.name)
  const solutionSchema: JSONSchemaType<Solution> = {
    type: 'array',
    items: {
      type: 'object',
      properties: {
        exe: {type: 'string', minLength: 1},
        sourceChain: {type: 'string', enum: chainName},
        destChain: {type: 'string', enum: chainName},
        spendAsset: {type: 'string', pattern: hexPattern},
        receiveAsset: {type: 'string', pattern: hexPattern},
      },
      required: [
        'exe',
        'sourceChain',
        'destChain',
        'spendAsset',
        'receiveAsset',
      ],
      additionalProperties: false,
    },
    minItems: 1,
  }

  const validate = ajv.compile(solutionSchema)

  return (solution: any): boolean => {
    let isValid = validate(solution)

    for (let i = 0; isValid && i < solution.length; i++) {
      const {sourceChain, destChain, spendAsset, receiveAsset} = solution[i]
      const isSwap = sourceChain === destChain
      if (
        // spendAsset and receiveAsset must be different when is swap
        (isSwap && spendAsset === receiveAsset) ||
        // sourceChain must be same with destChain of previous step
        (i > 0 && sourceChain !== solution[i - 1].destChain)
      ) {
        isValid = false
      }
    }
    return isValid
  }
}
