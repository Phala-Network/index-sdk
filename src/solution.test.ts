import {expect, test} from 'vitest'
import {createValidateFn} from './solution.js'

const chains: Chain[] = [
  {
    id: 0,
    name: 'Ethereum',
    endpoint: 'https://mainnet.infura.io/v3/6d61e7957c1c489ea8141e947447405b',
    chainType: 'Evm',
    nativeAsset: '0x0000000000000000000000000000000000000000',
    foreignAsset: null,
    handlerContract: '0xf9eae3ec6bfe94f510eb3a5de8ac9deb9e74df39',
    txIndexer: 'null',
  },
  {
    id: 1,
    name: 'Moonbeam',
    endpoint: 'https://moonbeam.api.onfinality.io/public',
    chainType: 'Evm',
    nativeAsset: '0x0000000000000000000000000000000000000000',
    foreignAsset: null,
    handlerContract: '0x635ea86804200f80c16ea8eddc3c749a54a9c37d',
    txIndexer: 'https://squid.subsquid.io/graph-moonbeam/graphql',
  },
  {
    id: 2,
    name: 'AstarEvm',
    endpoint: 'https://astar.public.blastapi.io',
    chainType: 'Evm',
    nativeAsset: '0x0000000000000000000000000000000000000000',
    foreignAsset: null,
    handlerContract: '0xb376b0ee6d8202721838e76376e81eec0e2fe864',
    txIndexer: 'https://squid.subsquid.io/graph-astar/graphql',
  },
  {
    id: 3,
    name: 'Astar',
    endpoint: 'https://astar.public.blastapi.io',
    chainType: 'Sub',
    nativeAsset: '0x010100591f',
    foreignAsset: 'PalletAsset',
    handlerContract: '0x00',
    txIndexer: 'https://squid.subsquid.io/graph-astar/graphql',
  },
  {
    id: 4,
    name: 'Khala',
    endpoint: 'https://khala-api.phala.network/rpc',
    chainType: 'Sub',
    nativeAsset: '0x0000',
    foreignAsset: 'PalletAsset',
    handlerContract: '0x79',
    txIndexer: 'https://squid.subsquid.io/graph-khala/graphql',
  },
  {
    id: 5,
    name: 'Phala',
    endpoint: 'https://api.phala.network/rpc',
    chainType: 'Sub',
    nativeAsset: '0x0000',
    foreignAsset: 'PalletAsset',
    handlerContract: '0x79',
    txIndexer: 'https://squid.subsquid.io/graph-phala/graphql',
  },
  {
    id: 6,
    name: 'Acala',
    endpoint: 'https://acala-rpc.dwellir.com',
    chainType: 'Sub',
    nativeAsset: '0x010200411f06080000',
    foreignAsset: 'OrmlToken',
    handlerContract: '0x00',
    txIndexer: 'https://squid.subsquid.io/graph-acala/graphql',
  },
]

const validate = createValidateFn(chains)

const cases: [any, boolean][] = [
  [
    [
      {
        exe_type: 'swap',
        exe: 'moonbeam_stellaswap',
        source_chain: 'Moonbeam',
        dest_chain: 'Moonbeam',
        spend_asset: '0xFfFFfFff1FcaCBd218EDc0EbA20Fc2308C778080',
        receive_asset: '0xFFFfFfFf63d24eCc8eB8a7b5D0803e900F7b6cED',
      },
      {
        exe_type: 'bridge',
        exe: 'moonbeam_bridge_to_phala',
        source_chain: 'Moonbeam',
        dest_chain: 'Phala',
        spend_asset: '0xFFFfFfFf63d24eCc8eB8a7b5D0803e900F7b6cED',
        receive_asset: '0x0000',
      },
    ],
    true,
  ],
  [
    [
      {
        exe_type: 'bridge_a',
        exe: 'moonbeam_bridge_to_phala',
        source_chain: 'Moonbeam_a',
        dest_chain: 'Phala_a',
        spend_asset: '0xFFFfFfFf63d24eCc8eB8a7b5D0803e900F7b6cED_a',
        receive_asset: '0x0000_a',
      },
    ],
    false,
  ],
  [
    [
      {
        exe_type: 'bridge',
        exe: 'moonbeam_bridge_to_phala',
        source_chain: 'Moonbeam',
        dest_chain: 'Moonbeam',
        spend_asset: '0xFFFfFfFf63d24eCc8eB8a7b5D0803e900F7b6cED',
        receive_asset: '0x0000',
      },
    ],
    false,
  ],
  [
    [
      {
        exe_type: 'swap',
        exe: 'moonbeam_bridge_to_phala',
        source_chain: 'Moonbeam',
        dest_chain: 'Moonbeam',
        spend_asset: '0xFFFfFfFf63d24eCc8eB8a7b5D0803e900F7b6cED',
        receive_asset: '0xFFFfFfFf63d24eCc8eB8a7b5D0803e900F7b6cED',
      },
    ],
    false,
  ],
  [
    [
      {
        exe_type: 'swap',
        exe: 'moonbeam_bridge_to_phala',
        source_chain: 'Moonbeam',
        dest_chain: 'Phala',
        spend_asset: '0xFFFfFfFf63d24eCc8eB8a7b5D0803e900F7b6cED',
        receive_asset: '0x0000',
      },
    ],
    false,
  ],
  [
    [
      {
        exe_type: 'swap',
        exe: 'moonbeam_stellaswap',
        source_chain: 'Moonbeam',
        dest_chain: 'Moonbeam',
        spend_asset: '0xFfFFfFff1FcaCBd218EDc0EbA20Fc2308C778080',
        receive_asset: '0xFFFfFfFf63d24eCc8eB8a7b5D0803e900F7b6cED',
      },
      {
        exe_type: 'bridge',
        exe: 'moonbeam_bridge_to_phala',
        source_chain: 'Khala',
        dest_chain: 'Phala',
        spend_asset: '0xFFFfFfFf63d24eCc8eB8a7b5D0803e900F7b6cED',
        receive_asset: '0x0000',
      },
    ],
    false,
  ],
]

for (let i = 0; i < cases.length; i++) {
  const [solution, valid] = cases[i]
  test(`Validate solution ${i + 1}`, () => {
    expect(validate(solution)).toBe(valid)
  })
}
