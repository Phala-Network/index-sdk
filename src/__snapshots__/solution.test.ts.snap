// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Solution > Validate solution 1 1`] = `true`;

exports[`Solution > Validate solution 1 2`] = `"0x0401084c6d6f6f6e6265616d5f7374656c6c6173776170204d6f6f6e6265616d204d6f6f6e6265616da8307846664646664666663146636143426432313845446330456241323046633233303843373738303830a8307846464666466646663633643234654363386542386137623544303830336539303046376236634544606d6f6f6e6265616d5f6272696467655f746f5f7068616c61204d6f6f6e6265616d145068616c61a830784646466646664666363364323465436338654238613762354430383033653930304637623663454418307830303030"`;

exports[`Solution > Validate solution 2 1`] = `true`;

exports[`Solution > Validate solution 2 2`] = `"0x10010c58657468657265756d5f6e61746976657772617070657220457468657265756d20457468657265756da8307830303030303030303030303030303030303030303030303030303030303030303030303030303030a830784330326161413339623232334645384430413065354334463237654144393038334337353643633248657468657265756d5f756e6973776170763220457468657265756d20457468657265756da8307843303261614133396232323346453844304130653543344632376541443930383343373536436332a830783663356241393136343246313032383262353736643931393232416536343438433964353266344574657468657265756d5f7379676d616272696467655f746f5f7068616c6120457468657265756d145068616c61a83078366335624139313634324631303238326235373664393139323241653634343843396435326634451830783030303000547068616c615f6272696467655f746f5f6173746172145068616c611441737461721830783030303030307830313031303063643166006061737461725f6272696467655f746f5f617374617265766d14417374617220417374617245766d30307830313031303063643166a830784646464646464646303030303030303030303030303030313030303030303030303030303030303601084861737461725f65766d5f617274687377617020417374617245766d20417374617245766da8307846464646464646463030303030303030303030303030303130303030303030303030303030303036a83078616561616630653263383161663236343130316239313239633030663434343063636630663732304861737461725f65766d5f617274687377617020417374617245766d20417374617245766da8307861656161663065326338316166323634313031623931323963303066343434306363663066373230a8307846464646464646463030303030303030303030303030303130303030303030303030303030303033"`;

exports[`Solution > Validate solution 3 1`] = `false`;

exports[`Solution > Validate solution 4 1`] = `false`;

exports[`Solution > Validate solution 5 1`] = `false`;
